### PR TITLE
Handler validator

### PR DESCRIPTION
**Pull request by allanmckenzie**

### CODE REVIEW

File: raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/RamlProvider.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import uk.gov.justice.raml.io.files.parser.RamlFileParser;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.Collection;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.raml.model.Raml;
+
+public class RamlProvider {
```
Comment (purple52) [inline raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/RamlProvider.java]: Again, why do we need this extra layer? It seems like a lot of extra code, just to wrap a single call to ramlOf. Why not just inject the RamlFileParser instead?

File: raml-maven-plugin/src/test/java/uk/gov/justice/raml/maven/lintchecker/LintCheckerMojoTest.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import uk.gov.justice.raml.maven.generator.BetterAbstractMojoTestCase;
+import uk.gov.justice.raml.maven.lintchecker.rules.FailingLintCheckRule;
+import uk.gov.justice.raml.maven.lintchecker.rules.SucceedingLintCheckRule;
+
+import java.io.File;
+
+import org.apache.maven.plugin.MojoExecutionException;
+import org.junit.Test;
+
+public class LintCheckerMojoTest extends BetterAbstractMojoTestCase {
```
Comment (purple52) [inline raml-maven-plugin/src/test/java/uk/gov/justice/raml/maven/lintchecker/LintCheckerMojoTest.java]: This doesn't test that the correct configuration/RAML gets passed to the rules.

File: raml-maven-plugin/src/test/resources/lint-check/pom.xml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+                <configuration>
+                    <sourceDirectory>${basedir}/src/raml</sourceDirectory>
+                    <includes>
+                        <include>**/*.raml</include>
+                    </includes>
+                    <excludes>
+                    </excludes>
+                    <rules>
+                        <MatchHandersActions implementation = "uk.gov.justice.raml.maven.lintchecker.rules.SucceedingLintCheckRule"/>
+                        <MatchHandersActions implementation = "uk.gov.justice.raml.maven.lintchecker.rules.FailingLintCheckRule"/>
```
Comment (purple52) [inline raml-maven-plugin/src/test/resources/lint-check/pom.xml]: Should these have different names?

File: raml-maven-plugin/src/test/resources/lint-check/src/raml/cakeshop-command-api.raml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#%RAML 0.8
```
Comment (purple52) [inline raml-maven-plugin/src/test/resources/lint-check/src/raml/cakeshop-command-api.raml]: Let's simplify this RAML (and JSON if its actually needed) to the bare minimum to its clear that we aren't testing the content.

File: lint-checker-core/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckRule.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package uk.gov.justice.raml.maven.lintchecker;
+
+import org.raml.model.Raml;
+
+/**
+ * Interface to be implemented by any rules executed by the lint checker.
+ */
+public interface LintCheckRule {
+
+    void execute(final Raml raml) throws LintCheckerException;
```
Comment (allanmckenzie) [inline lint-checker-core/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckRule.java]: added now

Comment (allanmckenzie) [inline lint-checker-core/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckRule.java]: fixed

File: lint-checker-core/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckerException.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package uk.gov.justice.raml.maven.lintchecker;
+
+public class LintCheckerException extends Exception {
```
Comment (allanmckenzie) [inline lint-checker-core/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckerException.java]: added now

File: raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/generator/GeneratorFactory.java [hunk 7]
```diff
@@ -22,22 +21,25 @@
...
 
-        }
+        try {
+            final Generator generator;
+            final Class<?> clazz = Class.forName(generatorName);
 
+            final Constructor<?> constructor = clazz.getConstructor();
+            generator = (Generator) constructor.newInstance();
+            return generator;
+        } catch (final Exception e) {
```
Comment (allanmckenzie) [inline raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/generator/GeneratorFactory.java]: fixed

File: raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckerMojoDelegate.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.List;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.MojoFailureException;
+import org.raml.model.Raml;
+
+public class LintCheckerMojoDelegate {
```
Comment (allanmckenzie) [inline raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckerMojoDelegate.java]: name changed to LintCheckGoalProcessor

File: raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckerMojoDelegate.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    private final RamlProvider ramlProvider;
+    private final PathsProvider pathsProvider;
+
+    public LintCheckerMojoDelegate() {
```
Comment (allanmckenzie) [inline raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckerMojoDelegate.java]: done

File: raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/PathsProvider.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.List;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.maven.plugin.MojoFailureException;
+
+public class PathsProvider {
```
Comment (allanmckenzie) [inline raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/PathsProvider.java]: changed to fileTreeScanner

File: raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/RamlProvider.java (hunk 1 above)
Comment (allanmckenzie) [inline raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/RamlProvider.java]: done

File: raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckerMojo.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.MojoFailureException;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.raml.model.Raml;
+
+
+@Mojo(name = "lint-check")
+public class LintCheckerMojo extends BasicMojo {
```
Comment (allanmckenzie) [inline raml-maven-plugin/src/main/java/uk/gov/justice/raml/maven/lintchecker/LintCheckerMojo.java]: done

File: raml-maven-plugin/src/test/resources/lint-check/src/raml/cakeshop-command-api.raml (hunk 4 above)
Comment (allanmckenzie) [inline raml-maven-plugin/src/test/resources/lint-check/src/raml/cakeshop-command-api.raml]: done

**Pull request merged**